sys: realtime: Add native real-time library and API

bjarki-andreasen opened the pull request:
## Introduction
This PR adds native APIs for setting, getting and working with real-time as unix timestamps and `struct tm` like datetimes.

## Overview
The following diagram presents an overview of the system realtime library. 
![sys-realtime drawio(2)](https://github.com/zephyrproject-rtos/zephyr/assets/82143143/6736e95e-fa0e-4e91-98c1-7b9304335361)
The system realtime clock, which is the base for `sys_realtime_get_/set_` APIs, is a settable complimentary clock to the sys_clock, which is the base for `k_uptime()`. The system realtime clock can be implemented using any high accuracy clock. This PR provides an implementation which uses the sys_clock as the base clock. Similarly, the implementation of the conversion functions `sys_realtime_timestamp_to_datetime` and opposite can be selected at build time. This PR provides a native implementation of said APIs.

## System realtime API
The top level API exposed to the user provides the following APIs
```
struct sys_datetime {
	int tm_sec;	/**< Seconds [0, 59] */
	int tm_min;	/**< Minutes [0, 59] */
	int tm_hour;	/**< Hours [0, 23] */
	int tm_mday;	/**< Day of the month [1, 31] */
	int tm_mon;	/**< Month [0, 11] */
	int tm_year;	/**< Year - 1900 */
	int tm_wday;	/**< Day of the week [0, 6] (Sunday = 0) (Unknown = -1) */
	int tm_yday;	/**< Day of the year [0, 365] (Unknown = -1) */
	int tm_isdst;	/**< Daylight saving time flag [-1] (Unknown = -1) */
	int tm_nsec;	/**< Nanoseconds [0, 999999999] (Unknown = 0) */
};

/** Get universal coordinated time unix timestamp in milliseconds */
__syscall int sys_realtime_get_timestamp(int64_t *timestamp_ms);
/** Set universal coordinated time unix timestamp in milliseconds */
__syscall int sys_realtime_set_timestamp(const int64_t *timestamp_ms);

/** Get universal coordinated time datetime */
__syscall int sys_realtime_get_datetime(struct sys_datetime *datetime);
/** Set universal coordinated time datetime */
__syscall int sys_realtime_set_datetime(const struct sys_datetime *datetime);

/** Convert universal coordinated time datetime to unix timestamp in milliseconds */
int sys_realtime_datetime_to_timestamp(int64_t *timestamp_ms, const struct sys_datetime *datetime);
/** Convert universal coordinated time unix timestamp in milliseconds to datetime */
int sys_realtime_timestamp_to_datetime(struct sys_datetime *datetime, const int64_t *timestamp_ms);
```
Thje get/set functions are syscalls, making them available to userspace, the conversion functions don't touch any kernel objects so they are not implemented as syscalls.

## Configuring initial time source
Setting realtime automatically on boot from a hardware RTC can be configured by using the chosen `zephyr,rtc`.
```
/ {
	chosen {
		zephyr,rtc = &rtc;
	};
};
```

fixes: #35333

bjarki-andreasen committed:
lib: sys: realtime: Add sys clock based realtime impl

Add system realtime clock implementation based on the system
clock (k_uptime)

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

fabiobaltieri commented on lib/time/convert.c:
```diff
@@ -0,0 +1,315 @@
+#include <zephyr/sys/time.h>
+#include <zephyr/sys/internal/time.h>
+
+#define SECONDS_IN_400_YEARS      12622780800
+#define SECONDS_IN_100_LEAP_YEARS 3155760000
+#define SECONDS_IN_100_YEARS      3155673600
+#define SECONDS_IN_4_YEARS        126144000
+#define SECONDS_IN_4_LEAP_YEARS   126230400
+#define SECONDS_IN_LEAP_YEAR      31622400
+#define SECONDS_IN_YEAR           31536000
+#define SECONDS_IN_DAY            86400
+#define SECONDS_IN_HOUR           3600
+#define SECONDS_IN_MINUTE         60
```
maybe express this in function of the previous one?

```c
define SECONDS_IN_HOUR (SECONDS_IN_MINUTE * 60)
```

and so on

bjarki-andreasen committed:
tests: lib: add sys_time conversion test suite

Add sys_time conversion test suite testing conversion between
and validation of datetimes and UTC timestamps, all the way from
year -1, to year 2399.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-trackunit commented:
## Update
### Added counter support
Using a low-power counter / real-time counter as timekeeper is now supported!
### Added test suite for time conversion functions
The test suite tests conversions from timestamps all the way back from year -1 to year 2399

fabiobaltieri commented on tests/lib/sys_time/convert/src/test_convert.c:
```diff
@@ -0,0 +1,217 @@
+#include <zephyr/sys/time.h>
+#include <string.h>
+
+struct test_sample
+{
```
brace goes up

bjarki-trackunit replied:
brace went up :)

bjarki-trackunit commented:
Note to self, namespace `SYS_TIME`

cfriedt reviewed (dismissed):
Looks ok to me. In terms of timestamp, is that seconds since the epoch?

bjarki-trackunit commented:
@cfriedt yep :) we could up the resolution to ms as well, we got the bits for it

fgrandel commented on include/zephyr/sys/internal/time.h:
```diff
@@ -0,0 +1,22 @@
+#include <zephyr/drivers/rtc.h>
+
+/** Validate timestamp can be converted to datetime */
+bool sys_time_validate_timestamp(const int64_t *timestamp);
```
Please call this `sys_rtc_time...` or any other `_rtc_` prefix or infix, as this API is closely  coupled to rtc.h. The time namespace is so crowded already in Zephyr that generic namespaces like sys_time should not be occupied by specific libraries.

fgrandel commented on lib/time/Kconfig:
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) 2024 Bjarki Arge Andreasen
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig TIME
```
Same here, the configuration should either be done in the RTC realm or refer to generic time concepts independent of a specific hardware type. This API is not general enough to represent general POSIX time.

bjarki-trackunit replied:
I will make an attempt to use this API as the basis for posix time, then we can discuss further. I will probably meet some of the hurdles you foresee :)

bjarki-trackunit commented on include/zephyr/sys/time.h:
```diff
@@ -0,0 +1,28 @@
+int sys_time_set_timestamp(const int64_t *timestamp);
+
+/** Get universal coordinated time datetime */
+int sys_time_get_datetime(struct rtc_time *datetime);
```
Its the other way around really :)

The RTC API itself is closely coupled to time.h, struct rtc_time is a 1-1 match to struct tm. The reason for this choice was to keep compatibility with POSIX, while not being dependent on it. This is the reason for using it over struct tm in the general API, not because of RTC drivers. We could move the struct out of rtc.h, have it in sys/time.h instead to make this clearer, actually, we should so that :)

cfriedt replied:
Oh yeah, I just realized it's using a standard header. It should really be moved to a separate one.

Generally speaking, we want to avoid adding non-standard functions to standard headers.

fgrandel commented on include/zephyr/sys/time.h:
```diff
@@ -0,0 +1,28 @@
+int sys_time_set_datetime(const struct rtc_time *datetime);
+
+/** Convert universal coordinated time datetime to timestamp */
+int sys_time_datetime_to_timestamp(int64_t *timestamp, const struct rtc_time *datetime);
```
Please make it clear that we're dealing with Unix timestamps here, as the timestamp name is also highly fraught in Zepyhr already.

bjarki-trackunit replied:
I will update the comments to specify unix time

fgrandel commented on lib/time/counter.c:
```diff
@@ -0,0 +1,305 @@
+	int ret;
+	uint32_t counter_ticks;
+	uint32_t counter_center_bottom_ticks;
+	uint32_t counter_center_top_ticks;
```
nit: reversed x-mas

bjarki-trackunit replied:
I will reverse the reversed name

fgrandel commented on lib/time/counter.c:
```diff
@@ -0,0 +1,305 @@
+
+static void add_ticks_to_seconds(int64_t *seconds, const uint64_t *ticks)
+{
+	*seconds += (int64_t)((*ticks) / frequency);
```
You seem to be making implicit rounding assumptions here, please document these assumptions and name the function accordingly. Also document your assumptions about overflow (if any), as well as any other concept relevant to counter-to-time-conversion (see the well-defined concepts and nomenclature in https://github.com/zephyrproject-rtos/zephyr/issues/19030#issuecomment-1597226731).

fgrandel commented on lib/time/counter.c:
counter != time-to-counter conversion - I suggest you name this properly as a conversion library (with very specific underlying assumptions) - again to not occupy generic namespaces with specific implementations.

bjarki-trackunit replied:
the name of the source file just needs to be specific enough to separate it from other files in the same folder, the name is not exposed anywhere outside the folder, so it is not occupying any generic namespace.

fgrandel replied:
Don't agree. We have been there and it has been highly confusing to users. We are speaking about highly visible header files. I vote for code that is not only "formally" correct but also helps users to not be trapped by misleading names.

fgrandel commented on lib/time/counter.c:
```diff
@@ -0,0 +1,305 @@
+	return sys_time_validate_timestamp(timestamp) ? 0 : -EAGAIN;
+}
+
+int sys_time_set_timestamp(const int64_t *timestamp)
```
The specific kind of counter and conversion assumptions should be decoupled from sys_time. Either make it clear where future decoupling can be done (even if not required now) or name the function more specifically and place it somewhere close to the specific type of counter API you're expecting here.

bjarki-trackunit replied:
I don't agree with this comment. The counter API has no relevance to the sys_time API, neither does the struct rtc_time have any relevance to the RTC API. Implementations of the sys_time API may use either, but the entire point is to abstract this away from the user, who simply wants to get/set and convert time.

I understand that the fact that we are reusing the `struct rtc_time` makes it seem like this is directly related to the rtc device driver API, it is not, `sys_time` works just fine on systems that don't have an RTC...

fgrandel reviewed (changes requested):
@andyross Would be great if you could give this a quick review wrt tick-to-counter conversion, duplication with other concepts and encapsulation/openness towards other system timepoint representations.

bjarki-trackunit commented:
How about naming it sys_realtime? Its a library handling interactions with and conversions between real-time clocks and real-time counters?

cfriedt commented:
It would be kind of great to have 1 header file for converting between zephyr time representations.

Naming things is hard, but if it can be added to an existing header, it's much easier.

fgrandel commented:
> It would be kind of great to have 1 header file for converting between zephyr time representations.
> 
> Naming things is hard, but if it can be added to an existing header, it's much easier.

I agree, that would be great, but the problem has been underestimated by many in the project already. There are lots of different ways to convert between pairs of time representations, especially when they come from different layers in the clock hierarchy as is the case here: counter layer vs. synchronized time layer - leaving out the syntonization layer in between w/o even mentioning it.

If you don't have good intermediate representations (single point of truth) on those three layers as well as APIs to manipulate them properly, then you get a combinatoric explosion of time types and conversion approaches. Hub and spoke is the only approach that works for proper time conversion IMHO.

Just think of leap seconds in the UTC clock and the large amount of established syntonization approaches across system borders which are both huge problems for long-running high-precision timing systems as those driven by Zephyr. Just to name two of those, but there are many more.

Due to our lack of proper time representation in Zephyr, we currently experience that kind of combinatoric explosion. This PR is just one example of many. Linux has gone a long way to converge towards proper intermediate time representations. But our challenge is even bigger, as we cannot sensibly let go of truely deterministic timing even in higher clock layers (syntonization, synchronization), something Linux mostly fails to deliver today.

Wrt this PR: If we cannot avoid another arbitrary experimental time conversion library, then I vote for keeping it somewhere down in one driver subsystem where it originates.

In any case I vote against another flawed approach of "general" time conversion clogging our common namespace,

fgrandel commented:
> How about naming it sys_realtime?

"Realtime" is still much too overloaded as a namespace. As long as you don't keep the promise of actually dealing with what "realtime" means in other systems, as long as we don't even reach the Linux benchmark for real time (much less deterministic synchronized clocks as required by several RTOS use cases), then this name promises far too much.

Let's face it: This is an RTC driver/hardware specific conversion library that might even solve an immediate problem in the POSIX layer for a little while but that will have to be replaced as soon as POSIX APIs want to expose something that is similar to what "real time" means on other systems.

I'm totally open to help you start with the nucleus of a well-defined time representation in Zephyr, but this PR is not the way to go to achieve this IMO. We have experimental implementations and specific APIs that go far beyond this PR already, so you'd first have to go beyond these.

cfriedt commented:
I would say, the basis for POSIX time is `struct timespec` rather than a 64-bit timestamp.

`struct rtctime` could be converted to a 64 bit ms counter since the epoch and then that could be converted into a `struct timespec` to set `CLOCK_REALTIME` for example.

That would provide a kind of coarse granularity "clock recovery" on power-on (in the absence of synchronizing via e.g. NTP or PTP).

Additionally, I think that would make a great `SYS_INIT()` feature of an rtc subsystem, along with shell utilities and so on. From that perspective, this code would be a consumer of the POSIX API.

Hard to say where to park this, but since it's not part of any standard, I would steer clear of `sys/time.h` or `time.h` since they are standard and kind of overloaded already.

andyross commented on lib/time/Kconfig:
```diff
@@ -0,0 +1,25 @@
+
+config TIME_MIN_YEAR
+	int "Minimum year to support for real-time services"
+	default 2000
```
Not really understanding why the defaults are so limited here.  The rtc_time struct is basically just a Unix struct tm and the year field is a 32 bit signed int with the zero at 1900.

andyross commented on lib/time/convert.c:
```diff
@@ -0,0 +1,334 @@
+	       datetime->tm_min >= 0 &&
+	       datetime->tm_min <= 59 &&
+	       datetime->tm_sec >= 0 &&
+	       datetime->tm_sec <= 59;
```
So, this isn't something called only when e.g. CONFIG_ASSERT=y, it's on the path to every call.  And this is actually a lot of code.  What is it trying to protect agaist?  If the user passes an incorrectly constructed date, it's IMHO perfectly legal to hand them back a nonsense timestamp.  Likewise if the RTC driver gives you a broken struct, it's not your problem to solve here.

andyross commented on lib/time/convert.c:
```diff
@@ -0,0 +1,334 @@
+	       datetime->tm_sec <= 59;
+}
+
+void sys_time_init_datetime(struct rtc_time *datetime)
```
API nitpick: why does this function exist?  What does it mean to "initialize" a date that doesn't actually (by construction!) refer to an actual date.  This literally produces an invalid struct per the validate() call above. ;)

As I see it this should just be removed.  You initialize a struct rtc_time by getting one from a driver or by converting a timestamp.

andyross reviewed (commented):
Answering the summons, though I'll note that while I know timing hardware and common traps with timer drivers and timing-sensitive app code, I'm really **not** much of an expert on Gregorian calendar stuff, except to say that I know it's historically difficult to get right.

Basically the worry here is that we're inventing a novel implementation of a library that the rest of the world has messed up repeatedly. :)

One thing that would be good to see is a host-side (or native_posix) unit test that exhaustively runs this and compares it against e.g. the glibc `gmtime()` implementation.  Like, pick random timestamps in the range of "all the times we'd expect this code to ever be used" and verify that it never differs over a few hundred billion iterations or whatever.

fgrandel reviewed (commented):
Returning to "comment" now that all arguments are on the table. Won't approve for the reasons given but also do not want to stand in the way of others who see merit in this contribution of course.

bjarki-andreasen committed:
lib: add realtime API and userspace handlers

Add user facing API for getting and setting system realtime and
converting between datetime (clock and calendar) and timestamp
(unixtime in milliseconds).

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

cfriedt commented:
What about "rtcutil" as a prefix instead of "sys_realtime"? Just a thought

fgrandel commented:
I propose `s/sys/rtc/g`, then place the API in rtc.h, see similar conversion APIs in other driver subsystems as counter and sensor. Let it mature there and expose it more prominently once it has proven useful to at least one more subsystem.

bjarki-andreasen committed:
lib: posix: clock: Refactor to use sys_realtime for realtime clock

The posix clock implementation uses k_uptime as its only source of
truth for time, tying clock monotonic 1-1 to k_uptime, and clock
realtime to (k_uptime + a reference/offset).

sys_realtime provides the zephyr version of a realtime clock, which
is managed by the kernel.

This commit refactors the posix clock implementation to offload the
realtime clock to sys_realtime, converting between a unix timestamp
in ms (sys_realtime) to a unix timestamp in struct timespec (posix
time).

A small adjustment to the posix common clock test suite to set the
realtime clock to the same time as clock monotonic was required
since the test suite can not support a clock which is not tied
1-1 to k_uptime.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
tests: posix: common: nanosleep: Update real-time test

The real time clock is almost never synced with the monotonic
clock, nor k_uptime which the monotonic clock is tied to.

Update the nanosleep test suite to set the real time clock
to a realistic value, apart from the monotonic clock. This
requires passing two new params to the common lower bound
check function, the requested time to sleep, which is now
different from the absolute k_uptime/monotonic clock time.

This update validates the the real time clock is settable
and indeed set to requested value, and that is running as
expected, and that the real time clock a different clock
than monotonic clock.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
drivers: rtc: shell: Add sys_realtime sync command

Add rtc shell command "sync" which synchronizes the system
realtime with the RTC time of the given RTC. The command
is used like:

    $uart: rtc sync rtc

to set the system realtime clock to the time of the rtc.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
lib: realtime: add native conversion library

Add native implementation of realtime conversion functions.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
tests: lib: sys_realtime: add real-time test suite

Add test suite which tests the timestamp and datetime get/set
APIs work from userspace and the implementation of the system
realtime clock is settable and incrementing correctly.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
drivers: rtc: add hwclock

The hwclock is selected using the chosen node "zephyr,rtc". It
will automatically set the system real-time clock on boot if the
RTCs time is valid.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
tests: test patch to hunt down infinite timeout in CI

Something is causing a test to sleep forever in CI, It can not
be reproduces locally, so a bunch of logs have been added to
try to see at least where it is failing.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

madscientist159 commented:
Is this still moving forward?  Right now we're having to do yet another roll-your-own solution to handle the fact that the only concept of time Zephyr has is its own uptime.  Uptime alone doesn't really work for network enabled applications or even for system event logging, and something better than an application-specific offset + uptime workaround would be greatly appreciated.

bjarki-andreasen commented:
Its quite dormant since I am spending most of my time with power management currently, I think the solution in this PR is nice, the only thing I remember needing to add was handling day of week and day of year conversions (if anyone wants that), would you like to help finalize this?

bjarki-andreasen commented:
oh, and also, there was a posix clock test which was failing for some target run in renode, 99% certain it was just some CI instability, but that was the last thing I was dealing with before switching tasks :)
